src/SkeletalTrapezoidation.cpp

```diff
@@ -0,0 +1,3083 @@
+    debugCheckGraphConsistency();
+    debugCheckGraphExistance();
+    debugCheckEndpointUniqueness();
+    
+    collapseSmallEdges();
+
+
+    debugCheckGraphCompleteness();
+    debugCheckGraphConsistency();
+    debugCheckGraphExistance();
+
+    { // set [some_edge] the the first possible edge
+        // that way we can iterate over all reachable edges from node.some_edge without needing to iterate backward
+        for (edge_t& edge : graph.edges)
+        {
+            if (!edge.prev)
+            {
+                edge.from->some_edge = &edge;
+            }
+        }
+    }
+
+
+    debugCheckGraphCompleteness();
+    debugCheckGraphConsistency();
+    debugCheckGraphStructure();
+    debugCheckGraphReachability();
+    debugCheckGraphExistance();
+    
+#ifdef DEBUG
+    {
+        AABB aabb(polys);
+        SVG svg("output/vq.svg", aabb);
+        debugOutput(svg, false, false);
+    }
+    {
+        AABB aabb(polys);
+        SVG svg("output/radial_dists.svg", aabb);
+        debugOutput(svg, false, true);
+    }
+    {
+        AABB aabb(polys);
+        SVG svg("output/bead_counts.svg", aabb);
+        debugOutput(svg, false, false, true);
+    }
+    {
+        AABB aabb(polys);
+        SVG svg("output/locations.svg", aabb);
+        debugOutput(svg, false, false, false, true);
+    }
+    debugCheckGraphCompleteness();
+    debugCheckGraphConsistency();
+
+    if (generate_MAT_STL)
+    {
+        STLwriter stl("output/mat.stl");
+        debugOutput(stl);
+    }
+#endif
+
+    vd_edge_to_he_edge.clear();
+    vd_node_to_he_node.clear();
+}
+
+void SkeletalTrapezoidation::separatePointyQuadEndNodes()
+{
+    std::unordered_set<node_t*> visited_nodes;
+    for (edge_t& edge : graph.edges)
+    {
+        if (edge.prev) continue;
+        edge_t* quad_start = &edge;
+        if (visited_nodes.find(quad_start->from) == visited_nodes.end())
+        {
+            visited_nodes.emplace(quad_start->from);
+        }
+        else
+        { // needs to be duplicated
+            graph.nodes.emplace_back(*quad_start->from);
+            node_t* new_node = &graph.nodes.back();
+            new_node->some_edge = quad_start;
+            quad_start->from = new_node;
+            quad_start->twin->to = new_node;
+        }
+    }
+
+    debugCheckEndpointUniqueness();
+}
+
+void SkeletalTrapezoidation::collapseSmallEdges(coord_t snap_dist)
+{
+    std::unordered_map<edge_t*, std::list<edge_t>::iterator> edge_locator;
+    std::unordered_map<node_t*, std::list<node_t>::iterator> node_locator;
+    for (auto edge_it = graph.edges.begin(); edge_it != graph.edges.end(); ++edge_it)
+        edge_locator.emplace(&*edge_it, edge_it);
+    for (auto node_it = graph.nodes.begin(); node_it != graph.nodes.end(); ++node_it )
+        node_locator.emplace(&*node_it, node_it );
+    auto safelyRemoveEdge = [this, &edge_locator](edge_t* to_be_removed, std::list<edge_t>::iterator& current_edge_it, bool& edge_it_is_updated)
```

Comment:
Like i said for the other comment, this one works on the `edge_locator` collection defined in the function rather than the graph itself.

Moving the `edge_locator` to the edge class is a bit overkill imho, since it only occurs within this function (the surrounding one, not the closure).